File: leetcode-problems/0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last = {}          # char -> last index
        left = 0           # left end of window
        best = 0

        for right, ch in enumerate(s):
            if ch in last and last[ch] >= left:
                left = last[ch] + 1      # move left past previous same char
            last[ch] = right
            best = max(best, right - left + 1)

        return best

File: leetcode-problems/0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> res;
        int n = s.length();
        int left = 0;
        int maxlen = 0;
        for(int right = 0; right< n; right++){
            if(res.count(s[right]) && res[s[right]] >= left){
                left = res[s[right]] + 1;
            }
            res[s[right]] =  right;
            maxlen = max(maxlen, right - left + 1);
            
        }
        return maxlen;
    }
};
